File: Raven.Client.Lightweight/Connection/MultiGetOperation.cs
using System;
using System.Collections.Specialized;
using System.Linq;
using System.Threading.Tasks;
using Raven.Abstractions;
using Raven.Client.Connection.Implementation;
using Raven.Client.Exceptions;
using Raven.Imports.Newtonsoft.Json;
using Raven.Abstractions.Data;
using Raven.Client.Connection.Profiling;
using Raven.Client.Document;
using Raven.Json.Linq;

namespace Raven.Client.Connection
{
    internal class MultiGetOperation
    {
        private readonly IHoldProfilingInformation holdProfilingInformation;
        private readonly DocumentConvention convention;
        private readonly string url;
        private readonly GetRequest[] requests;
        private readonly string requestUri;
        private bool allRequestsCanBeServedFromAggressiveCache;
        private CachedRequest[] cachedData;

        public string RequestUri
        {
            get { return requestUri; }
        }

        public MultiGetOperation(
            IHoldProfilingInformation holdProfilingInformation,
            DocumentConvention convention,
            string url,
            GetRequest[] requests)
        {
            this.holdProfilingInformation = holdProfilingInformation;
            this.convention = convention;
            this.url = url;
            this.requests = requests;

            requestUri = url + "/multi_get";
            if (convention.UseParallelMultiGet)
            {
                requestUri += "?parallel=yes";
            }
        }

        public GetRequest[] PreparingForCachingRequest(HttpJsonRequestFactory jsonRequestFactory)
        {
            cachedData = new CachedRequest[requests.Length];
            var requestsForServer = new GetRequest[requests.Length];
            Array.Copy(requests, 0, requestsForServer, 0, requests.Length);
            if (jsonRequestFactory.DisableHttpCaching == false && convention.ShouldCacheRequest(requestUri))
            {
                for (int i = 0; i < requests.Length; i++)
                {
                    var request = requests[i];
                    var cachingConfiguration = jsonRequestFactory.ConfigureCaching(url + request.UrlAndQuery,
                                                                                   (key, val) => request.Headers[key] = val);
                    cachedData[i] = cachingConfiguration.CachedRequest;
                    if (cachingConfiguration.SkipServerCheck)
                        requestsForServer[i] = null;
                }
            }
            allRequestsCanBeServedFromAggressiveCache = requestsForServer.All(x => x == null);
            return requestsForServer;
        }

        public bool CanFullyCache(HttpJsonRequestFactory jsonRequestFactory, HttpJsonRequest httpJsonRequest, string postedData)
        {
            if (allRequestsCanBeServedFromAggressiveCache) // can be fully served from aggressive cache
            {
                if (jsonRequestFactory.CanLogRequest)
                {

                    jsonRequestFactory.OnLogRequest(holdProfilingInformation, new RequestResultArgs
                    {
                        DurationMilliseconds = httpJsonRequest.CalculateDuration(),
                        Method = httpJsonRequest.Method,
                        HttpResult = 0,
                        Status = RequestStatus.AggressivelyCached,
                        Result = "",
                        Url = httpJsonRequest.Url.ToString(),
                        //TODO: check that is the same as: Url = httpJsonRequest.webRequest.RequestUri.PathAndQuery,
                        PostedData = postedData
                    });
                }
                return true;
            }
            return false;
        }

        public GetResponse[] HandleCachingResponse(GetResponse[] responses, HttpJsonRequestFactory jsonRequestFactory)
        {
            var hasCachedRequests = false;
            var requestStatuses = new RequestStatus[responses.Length];
            for (int i = 0; i < responses.Length; i++)
            {
                if (responses[i] == null || responses[i].Status == 304)
                {
                    hasCachedRequests = true;

                    requestStatuses[i] = responses[i] == null ? RequestStatus.AggressivelyCached : RequestStatus.Cached;
                    responses[i] = responses[i] ?? new GetResponse { Status = 0 };

                    foreach (string header in cachedData[i].Headers)
                    {
                        cachedData[i].Time = SystemTime.UtcNow;
                        responses[i].Headers[header] = cachedData[i].Headers[header];
                    }
                    responses[i].Result = cachedData[i].Data.CloneToken();
                    jsonRequestFactory.IncrementCachedRequests();
                }
                else
                {
                    requestStatuses[i] = responses[i].RequestHasErrors() ? RequestStatus.ErrorOnServer : RequestStatus.SentToServer;

                    var nameValueCollection = new NameValueCollection();
                    foreach (var header in responses[i].Headers)
                    {
                        nameValueCollection[header.Key] = header.Value;
                    }
                    jsonRequestFactory.CacheResponse(url + requests[i].UrlAndQuery, responses[i].Result, nameValueCollection);
                }
            }

            if (hasCachedRequests == false || convention.DisableProfiling ||
                holdProfilingInformation.ProfilingInformation.Requests.Count == 0)
                return responses;

            var lastRequest = holdProfilingInformation.ProfilingInformation.Requests.Last();
            for (int i = 0; i < requestStatuses.Length; i++)
            {
                lastRequest.AdditionalInformation["NestedRequestStatus-" + i] = requestStatuses[i].ToString();
            }
            lastRequest.Result = JsonConvert.SerializeObject(responses);

            return responses;
        }

        public async Task TryResolveConflictOrCreateConcurrencyException(GetResponse[] responses, Func<string, RavenJObject, Etag, Task<ConflictException>> tryResolveConflictOrCreateConcurrencyException)
        {
            // ReSharper disable once ForCanBeConvertedToForeach
            // see: http://ayende.com/blog/169377/the-bug-that-ruined-my-weekend
            for (int index = 0; index < responses.Length; index++)
            {
                var response = responses[index];
                if (response == null)
                    continue;
                if (response.RequestHasErrors() && response.Status != 409)
                    continue;

                var result = response.Result as RavenJObject;
                if (result == null)
                    continue;

                if (result.ContainsKey("Results"))
                {
                    var results = result["Results"] as RavenJArray;
                    if (results == null)
                        continue;

                    foreach (RavenJToken value in results)
                    {
                        var docResult = value as RavenJObject;
                        if (docResult == null)
                            return;

                        var metadata = docResult[Constants.Metadata];
                        if (metadata == null)
                            return;

                        if (metadata.Value<int>("@Http-Status-Code") != 409)
                            return;

                        var id = metadata.Value<string>("@id");
                        var etag = HttpExtensions.EtagHeaderToEtag(metadata.Value<string>("@etag"));

                        await TryResolveConflictOrCreateConcurrencyExceptionForSingleDocument(
                            tryResolveConflictOrCreateConcurrencyException,
                            id,
                            etag,
                            docResult,
                            response)
                            .ConfigureAwait(false);
                    }

                    continue;
                }
                if (result.ContainsKey("Conflicts"))
                {
                    var id = response.Headers[Constants.DocumentIdFieldName];
                    var etag = response.GetEtagHeader();

                    await TryResolveConflictOrCreateConcurrencyExceptionForSingleDocument(
                        tryResolveConflictOrCreateConcurrencyException,
                        id,
                        etag,
                        result,
                        response)
                        .ConfigureAwait(false);
                }
            }
        }

        private static async Task TryResolveConflictOrCreateConcurrencyExceptionForSingleDocument(
            Func<string, RavenJObject, Etag, Task<ConflictException>> tryResolveConflictOrCreateConcurrencyException,
            string id,
            Etag etag,
            RavenJObject docResult,
            GetResponse response)
        {
            var concurrencyException = await tryResolveConflictOrCreateConcurrencyException(id, docResult, etag).ConfigureAwait(false);

            if (concurrencyException != null)
                throw concurrencyException;

            response.Status = 200;
            response.ForceRetry = true;
        }
    }
}

File: Raven.Client.Java/src/main/java/net/ravendb/client/connection/MultiGetOperation.java
package net.ravendb.client.connection;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import net.ravendb.abstractions.closure.Action2;
import net.ravendb.abstractions.data.GetRequest;
import net.ravendb.abstractions.data.GetResponse;
import net.ravendb.client.connection.implementation.HttpJsonRequest;
import net.ravendb.client.connection.implementation.HttpJsonRequestFactory;
import net.ravendb.client.connection.profiling.IHoldProfilingInformation;
import net.ravendb.client.connection.profiling.RequestResultArgs;
import net.ravendb.client.connection.profiling.RequestStatus;
import net.ravendb.client.document.DocumentConvention;
import net.ravendb.imports.json.JsonConvert;

import org.apache.http.HttpStatus;


public class MultiGetOperation {
  private IHoldProfilingInformation holdProfilingInformation;
  private DocumentConvention convention;
  private String url;
  private GetRequest[] requests;
  private String requestUri;
  private boolean allRequestsCanBeServedFromAggressiveCache;
  private CachedRequest[] cachedData;

  public String getRequestUri() {
    return requestUri;
  }

  public MultiGetOperation(IHoldProfilingInformation holdProfilingInformation, DocumentConvention convention, String url, GetRequest[] requests) {
    this.holdProfilingInformation = holdProfilingInformation;
    this.convention = convention;
    this.url = url;
    this.requests = requests;

    requestUri = url + "/multi_get";
    if (convention.isUseParallelMultiGet())
    {
      requestUri += "?parallel=yes";
    }
  }

  private static class SetHeaders implements Action2<String, String> {
    private GetRequest getRequest;

    public SetHeaders(GetRequest getRequest) {
      this.getRequest = getRequest;
    }

    @Override
    public void apply(String headerName, String value) {
      getRequest.getHeaders().put(headerName, value);
    }

  }

  public GetRequest[] preparingForCachingRequest(HttpJsonRequestFactory jsonRequestFactory) {
    cachedData = new CachedRequest[requests.length];
    GetRequest[] requestsForServer = Arrays.copyOf(requests, requests.length);
    if (jsonRequestFactory.getDisableHttpCaching() == false && convention.shouldCacheRequest(requestUri)) {
      for (int i = 0; i < requests.length; i++) {
        GetRequest request = requests[i];
        CachedRequestOp cachingConfiguration = jsonRequestFactory.configureCaching(url + request.getUrlAndQuery(), new SetHeaders(request));
        cachedData[i] = cachingConfiguration.getCachedRequest();
        if (cachingConfiguration.isSkipServerCheck())
          requestsForServer[i] = null;
      }
    }
    boolean allNull = true;
    for(GetRequest request : requestsForServer) {
      allNull &= request == null;
    }
    allRequestsCanBeServedFromAggressiveCache = allNull;
    return requestsForServer;
  }

  public boolean canFullyCache(HttpJsonRequestFactory jsonRequestFactory, HttpJsonRequest httpJsonRequest, String postedData)
  {
    if (allRequestsCanBeServedFromAggressiveCache) { // can be fully served from aggressive cache
      RequestResultArgs args = new RequestResultArgs();

      args.setDurationMilliseconds(httpJsonRequest.calculateDuration());
      args.setMethod(httpJsonRequest.getMethod());
      args.setHttpResult(0);
      args.setStatus(RequestStatus.AGGRESSIVELY_CACHED);
      args.setResult("");
      args.setUrl(httpJsonRequest.getUrl());
      args.setPostedData(postedData);

      jsonRequestFactory.invokeLogRequest(holdProfilingInformation, args);
      return true;
    }
    return false;
  }

  public GetResponse[] handleCachingResponse(GetResponse[] responses, HttpJsonRequestFactory jsonRequestFactory) {
    boolean hasCachedRequests = false;
    RequestStatus[] requestStatuses = new RequestStatus[responses.length];
    for (int i = 0; i < responses.length; i++) {
      if (responses[i] == null || responses[i].getStatus() == HttpStatus.SC_NOT_MODIFIED) {
        hasCachedRequests = true;

        requestStatuses[i] = responses[i] == null ? RequestStatus.AGGRESSIVELY_CACHED : RequestStatus.CACHED;
        if (responses[i] == null) {
          responses[i] = new GetResponse();
          responses[i].setStatus(0);
        }

        for (String header: cachedData[i].getHeaders().keySet()) {
          responses[i].getHeaders().put(header, cachedData[i].getHeaders().get(header));
        }
        responses[i].setResult(cachedData[i].getData().cloneToken());
        jsonRequestFactory.incrementCachedRequests();
      } else {
        requestStatuses[i] = responses[i].isRequestHasErrors() ? RequestStatus.ERROR_ON_SERVER : RequestStatus.SEND_TO_SERVER;

        Map<String, String> nameValueCollection = new HashMap<>();
        for (Map.Entry<String, String> header: responses[i].getHeaders().entrySet()) {
          nameValueCollection.put(header.getKey(), header.getValue());
        }
        jsonRequestFactory.cacheResponse(url + requests[i].getUrlAndQuery(), responses[i].getResult(), nameValueCollection);
      }
    }

    if (hasCachedRequests == false || convention.isDisableProfiling() || holdProfilingInformation.getProfilingInformation().getRequests().size() == 0)
      return responses;

    List<RequestResultArgs> profilingRequests = holdProfilingInformation.getProfilingInformation().getRequests();
    RequestResultArgs lastRequest = profilingRequests.get(profilingRequests.size() - 1);
    for (int i = 0; i < requestStatuses.length; i++) {
      lastRequest.getAdditionalInformation().put("NestedRequestStatus-" + i, requestStatuses[i].toString());
    }
    lastRequest.setResult(JsonConvert.serializeObject(responses));

    return responses;
  }

}
